stratum/tools/benchmark/p4runtime_session.cc

```diff
@@ -0,0 +1,223 @@
+  *arbitration->mutable_election_id() = session->election_id_;
+  if (!session->stream_channel_->Write(request)) {
+    RETURN_ERROR(ERR_UNAVAILABLE)
+        << "Unable to initiate P4RT connection to device ID " << device_id
+        << "; gRPC stream channel closed.";
+  }
+
+  // Wait for arbitration response.
+  p4::v1::StreamMessageResponse response;
+  if (!session->stream_channel_->Read(&response)) {
+    RETURN_ERROR(ERR_INTERNAL)
+        << "No arbitration response received because: "
+        << GrpcStatusToStatus(session->stream_channel_->Finish())
+        << " with response: " << response.ShortDebugString();
+  }
+  if (response.update_case() != p4::v1::StreamMessageResponse::kArbitration) {
+    RETURN_ERROR(ERR_INTERNAL)
+        << "No arbitration update received but received the update of "
+        << response.update_case() << ": " << response.ShortDebugString();
+  }
+  if (response.arbitration().device_id() != session->device_id_) {
+    RETURN_ERROR(ERR_INTERNAL)
+        << "Received device id doesn't match: " << response.ShortDebugString();
+  }
+  if (response.arbitration().election_id().high() !=
+      session->election_id_.high()) {
+    RETURN_ERROR(ERR_INTERNAL)
+        << "Highest 64 bits of received election id doesn't match: "
+        << response.ShortDebugString();
+  }
+  if (response.arbitration().election_id().low() !=
+      session->election_id_.low()) {
+    RETURN_ERROR(ERR_INTERNAL)
+        << "Lowest 64 bits of received election id doesn't match: "
+        << response.ShortDebugString();
+  }
+
+  // Move is needed to make the older compiler happy.
+  // See: go/totw/labs/should-i-return-std-move.
+  return std::move(session);
+}
+
+// Creates a session with the switch, which lasts until the session object is
+// destructed.
+::util::StatusOr<std::unique_ptr<P4RuntimeSession>> P4RuntimeSession::Create(
+    const std::string& address,
+    const std::shared_ptr<grpc::ChannelCredentials>& credentials,
+    uint32_t device_id, absl::uint128 election_id) {
+  return Create(CreateP4RuntimeStub(address, credentials), device_id,
+                election_id);
+}
+
+// Create the default session with the switch.
+std::unique_ptr<P4RuntimeSession> P4RuntimeSession::Default(
+    std::unique_ptr<P4Runtime::Stub> stub, uint32_t device_id) {
+  // Using `new` to access a private constructor.
+  return absl::WrapUnique(
+      new P4RuntimeSession(device_id, std::move(stub), device_id));
+}
+
+::util::StatusOr<ReadResponse> SendReadRequest(
+    P4RuntimeSession* session, const ReadRequest& read_request) {
+  grpc::ClientContext context;
+  auto reader = session->Stub().Read(&context, read_request);
+
+  ReadResponse response;
+  ReadResponse partial_response;
+  while (reader->Read(&partial_response)) {
+    response.MergeFrom(partial_response);
+  }
+
+  grpc::Status reader_status = reader->Finish();
+  if (!reader_status.ok()) {
+    return GrpcStatusToStatus(reader_status);
+  }
```

Comment:
@ccascone this is how to handle batched wildcard reads spread over many responses, I believe.